autonomy/chain/mint.py

```diff
@@ -0,0 +1,159 @@
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Helpers for minting components"""
+
+import json
+from collections import OrderedDict
+from pathlib import Path
+from typing import Any, Dict, List, Optional, cast
+
+from aea.configurations.constants import DEFAULT_README_FILE
+from aea.configurations.data_types import PublicId
+from aea.crypto.base import Crypto, LedgerApi
+from aea.helpers.cid import CID
+from aea.helpers.ipfs.base import IPFSHashOnly
+from aea_cli_ipfs.ipfs_utils import IPFSTool
+from requests.exceptions import ConnectionError as RequestsConnectionError
+
+from autonomy.chain.base import ComponentRegistry, RegistriesManager
+from autonomy.chain.config import ChainType
+from autonomy.chain.exceptions import ComponentMintFailed, FailedToRetrieveTokenId
+
+
+BASE16_HASH_PREFIX = "f01701220"
+CONFIG_HASH_STRING_PREFIX = "0x"
+UNIT_HASH_PREFIX = CONFIG_HASH_STRING_PREFIX + "{metadata_hash}"
+DEFAULT_NFT_IMAGE_HASH = "bafybeiggnad44tftcrenycru2qtyqnripfzitv5yume4szbkl33vfd4abm"
+
+ContractInterfaceType = Any
+
+
+def serialize_metadata(
+    package_hash: str,
+    public_id: PublicId,
+    description: str,
+    nft_image_hash: str,
+) -> str:
+    """Serialize metadata."""
+    metadata = OrderedDict(
+        {
+            "name": f"{public_id.author}/{public_id.name}",
+            "description": description,
+            "code_uri": f"ipfs://{package_hash}",
+            "image": f"ipfs://{nft_image_hash}",
+            "attributes": [{"trait_type": "version", "value": str(public_id.version)}],
+        }
+    )
+    return json.dumps(obj=metadata)
+
+
+def publish_metadata(
+    public_id: PublicId,
+    package_path: Path,
+    nft_image_hash: str,
+) -> str:
+    """Publish service metadata."""
+
+    ipfs_tool = IPFSTool()
+    package_hash = IPFSHashOnly.get(file_path=str(package_path))
+    metadata_string = serialize_metadata(
+        package_hash=package_hash,
+        public_id=public_id,
+        description=Path(package_path, DEFAULT_README_FILE).read_text(encoding="utf-8"),
+        nft_image_hash=nft_image_hash,
+    )
+
+    metadata_hash = ipfs_tool.client.add_str(metadata_string)
+    metadata_hash = (
+        CID.from_string(metadata_hash)
+        .to_v1()
+        .encode("base16")
+        .decode()
+        .replace(BASE16_HASH_PREFIX, "")
+    )
+
+    return UNIT_HASH_PREFIX.format(metadata_hash=metadata_hash)
+
+
+def verify_and_fetch_token_id_from_event(
+    event: Dict,
+    unit_type: RegistriesManager.UnitType,
+    metadata_hash: str,
+    ledger_api: LedgerApi,
+) -> Optional[int]:
```

Comment:
After the token has been minted we wait for the relevant event to be emitted so we can extract the `unitId/tokenId` for the minted component